### PR TITLE
Simplestreams: Allow injection of global http client.

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following PR starts to extract the http client from the
bottom of the simplestreams code and lifts it up into a factory.
The factory can then be changed during testing to work against
local requests.

This then removes test code that is exercised in production, to
help facilitate patching of tests.

Unfortunately, as you can see from the PR, the simplestreams code
is in a bit of a state. Functions are not composed but instead are
called inside each other preventing correct testing. Testing was done
by patching functions out at various levels (std lib HTTP clients and
transports!) and canned returned values.

This is not how modeling of dependencies should happen.

## See

 - https://github.com/juju/juju/pull/13023

## QA steps

```sh
(cd tests && ./main.sh upgrade)
```

## Bug reference

 - https://bugs.launchpad.net/juju/+bug/1928182
 - https://bugs.launchpad.net/juju/+bug/1899793

### CODE REVIEW

File: environs/simplestreams/testing/testing.go [hunk 1]
```diff
@@ -547,11 +547,6 @@ var imageData = map[string]string{
 
 var TestRoundTripper = &testing.ProxyRoundTripper{}
 
-func init() {
```
Comment (SimonRichardson) [inline environs/simplestreams/testing/testing.go:550]: One down, only 68 left.

File: container/kvm/kvm.go [hunk 2]
```diff
@@ -26,14 +26,19 @@ import (
...
 var (
 	logger = loggo.GetLogger("juju.container.kvm")
 
-	// KvmObjectFactory implements the container factory interface for kvm
+	// KVMObjectFactory implements the container factory interface for kvm
 	// containers.
-	KvmObjectFactory ContainerFactory = &containerFactory{}
+	// TODO (stickupkid): This _only_ exists here because we can patch it in
+	// tests. This is horrid!
+	KVMObjectFactory ContainerFactory = &containerFactory{
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply container/kvm/kvm.go:39]: I tried that, but it means that we have to re-write all the kvm tests, which blows the PR out of the water.

[the remark this answers is not in the document]
Comment (SimonRichardson) [reply container/kvm/kvm.go:39]: See: https://github.com/juju/juju/blob/develop/container/kvm/testing/test.go#L35

**Review by achilleasa — APPROVED** [review]

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**